### DISCUSSION TITLE
latex table inserted with before-body.tex partial results in ERROR if there no other tables in the main body (PDF)

### DISCUSSION THREAD

**Discussion by kompre** [quarto-dev/quarto-cli#8808] [Q&A] [answered]
### Description

I'm trying to insert a revision table before the toc, but latex will fail to compile with an obscure error if there are no other tables in the qmd file (the main body). If a table is present in the qmd, then it compiles fine.

- the error message:

```
ERROR: 
compilation failed- error
Emergency stop.
<*> before-body-table.tex
```

I've got the latex syntax table from the tex file generated by quarto when rendering the qmd file containg the table, then copied to the `before-body.tex` partial, appending it at the end:

- the qmd that works:

```qmd
---
title: "before-body-table"
format: pdf
author: kompre
keep-tex: true
template-partials: 
    - before-body.tex
#    - tables.tex  # Edit: non necessary to reproduce the issue
toc: true
---

## Quarto

just comment the table out to get the error

|  Rev  | Descrizione |    Data    | Autore |
| :---: | ----------- | :--------: | :----: |
|   0   | Emissione   | 2024-02-15 |  FLL   |

```

- the partial `before-body.tex`:

```latex
$if(has-frontmatter)$
\frontmatter
$endif$
$if(title)$
$if(beamer)$
\frame{\titlepage}
$else$
\maketitle
$endif$
$if(abstract)$
\begin{abstract}
$abstract$
\end{abstract}
$endif$
$endif$

% modified from here
\begin{longtable}[]{@{}clcc@{}}
    \toprule\noalign{}
    Rev & Descrizione & Data & Autore \\
    \midrule\noalign{}
    \endhead
    \bottomrule\noalign{}
    \endlastfoot
    0 & Emissione & 2024-02-15 & FLL \\
    \end{longtable}
```

With a table in the qmd, the pdf is able to render correctly:

![image](https://github.com/quarto-dev/quarto-cli/assets/11796345/04698a08-7017-46e9-9791-8803622ae0cc)

But  if I comment out the table in the qmd file, meaning no table is present in the qmd file, then the error is raised.

I thought this could be because the `tables.tex` partial would not be properly loaded since no table were found in the text, but I checked the `tex` file and the relevant parts were loaded in the preamble in both cases.

Comparing the `tex` file between the working and failed render, I can see that they differs only after the `\maketitle`: the `tex` file of the failed render just stops after `\maketitle`, but the preamble is the same.

Lastly I changed the table in the `before-body-tex` to a different, simpler environment and it works:

```
\begin{tabular}{|c|l|c|c|}
    \hline
    Rev & Descrizione & Data       & Autore \\ \hline
    0   & Emissione   & 2024-02-15 & FLL    \\ \hline
    \end{tabular}%
```

What is happening with the `longtable `to cause shenanigans?

**Comment by mcanouil** [comment]
I'am afraid your example is not reproducible, *i.e.*, `tables.tex` does not exist.
Please ensure your example is fully reproducible and possibly share as a Git repository if there are multiple files.

I cannot reproduce the issue.
Make sure your code is correctly formatted.
<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/a0e866ce-9c1b-4fa5-a1f2-451d0279adbb">

**Reply by kompre** [reply]
I've updated the question commenting out the requirement for `tables.tex` because it not necessary to reproduce the issue (it's the left over of an earlier version of the question where I thought that partial was not loading due to any tables being present in the text, but I was mistaken)

**Comment by cderv** [comment, marked as answer]
Can you try indenting correctly the end of the environment in `before-body.tex` ? 

````latex
% modified from here
\begin{longtable}[]{@{}clcc@{}}
    \toprule\noalign{}
    Rev & Descrizione & Data & Autore \\
    \midrule\noalign{}
    \endhead
    \bottomrule\noalign{}
    \endlastfoot
    0 & Emissione & 2024-02-15 & FLL \\
\end{longtable}
````

Somehow this seems to mess up the partials feature from Pandoc. I can't explain yet, but when I try your example (without `tables.tex`) this is what I noticed.

**Reply by cderv** [reply]
In fact, I recall we have a lot of processing happening in Quarto on the produced `.tex` file, and we do expect `\end{environment}` to be first on line. So it matters to Quarto that the LaTeX code is properly formatted.

I think our processing happens on `longtable` not `tabular` and why you did not get the issue with `tabular`

Is there a reason to indent the closing of the environment with `\end{longtable}` prepend with spaces ?

**Reply by kompre** [reply]
Well, there is a reason, not a good one though: my process was to first generate the correct latex code with quarto then copy and paste the snippet of latex code in the `before-body.tex` and then change it. The space is added when I paste the latex code into the file, probably because some formatter of vscode.

Still, if a table is present in the body, the table included i the `before-body.tex` will render even if the `\end{environment}` is not properly aligned.
